taichi_3d_gaussian_splatting/utils.py

```diff
@@ -247,27 +253,36 @@ def get_point_probability_density_from_2d_gaussian_normalized(
     exponent = -0.5 * xy_mean_T_cov_inv_xy_mean
     return ti.exp(exponent)
 
+
 @ti.func
-def get_point_conic(
+def get_point_conic_and_rescale(
     gaussian_covariance: ti.math.mat2,
-) -> ti.math.vec3:
+) -> ti.math.vec4:
+    # Apply low-pass filter: every Gaussian should be at least one pixel wide/high. Discard 3rd row and column.
+    # https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/59f5f77e3ddbac3ed9db93ec2cfe99ed6c5d121d/cuda_rasterizer/forward.cu#L108-L111
+    det_cov_prefilter = gaussian_covariance.determinant()
+    gaussian_covariance[0, 0] += 0.3
+    gaussian_covariance[1, 1] += 0.3
     det_cov = gaussian_covariance.determinant()
-    inv_cov = (1. / det_cov) * \
+    rescale = ti.math.sqrt(ti.max(0.0, det_cov_prefilter / det_cov))
+    inv_cov = (1.0 / det_cov) * \
         ti.math.mat2([[gaussian_covariance[1, 1], -gaussian_covariance[0, 1]],
                       [-gaussian_covariance[1, 0], gaussian_covariance[0, 0]]])
-    conic = ti.math.vec3([inv_cov[0, 0], inv_cov[0, 1], inv_cov[1, 1]])
-    return conic
+    conic_and_rescale = ti.math.vec4([inv_cov[0, 0], inv_cov[0, 1], inv_cov[1, 1]], rescale)
```

Comment:
@wanmeihuali This is the main difference from official GS, where we calculate the rescale factor what will be later multiplied to alpha.

Reply:
I see, so the official implementation convolves each Gaussian with an isotropic 2D Gaussian to ensure each Gaussian is larger than a pixel, however, such operation changes the total integrated opacity of the Gaussian. So you normalize the opacity to ensure the total opacity does not change. It looks reasonable, but I still didn't fully get the connection between the issue and grid patterns... @jb-ye can you explain more about it? Thanks!

Reply:
In the grid pattern, white and brown gaussians form a net structure. particularly white Gaussians are formed behind the brown ones. In similar distance as captured cameras, they together render pixels in similar values as captured image. If view them in further distance, brown ones stand out because of the 2D convolution. So the pixels become more brownish when viewing from far distance.

Reply:
Got it! That's pretty reasonable, Thanks!